### PR TITLE
tickets/DM-42034: Inject HOMEDIR into init containers

### PR DESCRIPTION
**Pull request by athornton**

### CODE REVIEW

**Comment by athornton** [comment]
Since I'm rewriting initdir in python I'm going to move it into the monorepo now, and make that part of this PR.  Renaming the env vars is fine.

File: inithome/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,80 @@
+[project]
+# https://packaging.python.org/en/latest/specifications/declaring-project-metadata/
+name = "rubin.nublado.inithome"
```
Comment (rra) [inline inithome/pyproject.toml:3]: There isn't any problem with this name, really, but I've been using shorter names for things like this that we'll never upload to PyPI and whose output artifact is a Docker image. So following the controller, I'd just call this inithome, personally, and save the typing. But I think either is fine.

File: inithome/src/rubin/nublado/inithome/provisioner.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    def _validate(self, uid: int, gid: int) -> None:
+        for item in (uid, gid):
+            if item == 0:
+                raise ValueError("Will not provision for UID/GID 0")
+            if item < 0:
+                raise ValueError("UID/GID must be positive")
+            if item > MAX_ID:
+                raise ValueError(f"UID/GID must be <= {MAX_ID}")
+
+    async def provision(self) -> None:
```
Comment (rra) [inline inithome/src/rubin/nublado/inithome/provisioner.py:29]: I think you can just make this sync and save the `asyncio.run`. I can't imagine any case where we'd use async functions in this code.

File: inithome/src/rubin/nublado/inithome/provisioner.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            if stat_results.st_uid != self.uid:
+                raise RuntimeError(
+                    f"{self.homedir} is owned by {stat_results.st_uid}, not "
+                    f"{self.uid}"
+                )
+            if stat_results.st_gid != self.gid:
+                raise RuntimeError(
+                    f"{self.homedir} is owned by group "
+                    f"{stat_results.st_gid}, not {self.gid}"
+                )
```
Comment (rra) [inline inithome/src/rubin/nublado/inithome/provisioner.py:57]: I'm tempted to just fix the permissions rather than throw a fatal error, although I suppose that doesn't make sense if there are files. Maybe check if the directory is empty, and if so, just fix the permissions?

Specifically, I'm worried about the case where someone runs Nublado without an init container first without realizing it's needed, Kubernetes creates the directory owned as root, and now the user is stuck and needs manual intervention. It would be nice if we could self-heal from that scenario if it won't cause other problems.

File: inithome/src/rubin/nublado/inithome/provisioner.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            if (stat_results.st_mode & 0o777) != 0o700:
+                logger = logging.getLogger(__name__)
+                logger.warning(
+                    f"{self.homedir} has strange permissions "
+                    f"{oct(stat_results.st_mode & 0o777)} rather than '0o700'"
+                )
+            return
+        # We need to create the directory
+        # Set umask so intermediate paths are created mode 0o700 as well
+        umask(0o077)
```
Comment (rra) [inline inithome/src/rubin/nublado/inithome/provisioner.py:69]: This will then break access to this directory when using NFS with root squash. Shouldn't we be using 0711 for the parent directories as a safer default mode instead and only use 0700 for the lowest-level directory?

File: inithome/src/rubin/nublado/inithome/provisioner.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        # Find out how many layers of directory we need to create, for the
+        # chown.
+        first_created = self.homedir
+        while True:
+            if first_created.parent.is_dir() or first_created == Path("/"):
+                break
+            first_created = first_created.parent
+        # Create the homedir (and path leading down to it if needed)
+        self.homedir.mkdir(parents=True)
```
Comment (rra) [inline inithome/src/rubin/nublado/inithome/provisioner.py:78]: This logic is hard to follow, partly because `first_created` is not correctly named (it's the parent directory of the first directory that was created). I think I'd instead build a list of parent directories that need to be created, mkdir and chown each of them individually in order with mode 0711, and then make the actual home directory with mode 0700, without using `parents=True` since it's clearer with parallelism between each mkdir and chown.

File: inithome/src/rubin/nublado/inithome/provisioner.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
+"""Provisioner for user home directories."""
+
+import asyncio
+import logging
+from os import chown, environ, getenv, umask
```
Comment (rra) [inline inithome/src/rubin/nublado/inithome/provisioner.py:5]: Nit, but these days I more commonly see `import os` and then call these as `os.chown`, `os.environ`, etc. That would make it a bit more consistent with our other code.

File: inithome/tests/conftest.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    "gsamsa": User(homedir=Path("/home/gsamsa"), uid=2247, gid=200),
+    "josephk": User(
+        homedir=Path("/home/j/josephk/nublado"), uid=63928, gid=63928
+    ),
+    "leni": User(homedir=Path("/home/leni"), uid=2000, gid=200),
+    "grubach": User(homedir=Path("/home/grubach"), uid=9942, gid=500),
+    "karl": User(homedir=Path("/home/karl"), uid=1088, gid=500),
+    "huld": User(homedir=Path("/home/huld"), uid=4346, gid=4346),
+    "burstner": User(homedir=Path("/home/burstner"), uid=7304, gid=7304),
+}
```
Comment (rra) [inline inithome/tests/conftest.py:28]: It's always tricky to figure out how to divide this stuff between fixtures and tests, and I don't always do this cleanly (and sometimes it's almost impossible to do this cleanly). But in general I find the pattern where the initial test data is defined in one file and then the tests assume that data and make assertions about it in a different file to be hard to read. I have to keep switching back and forth between the files to understand the relationship between the test and the setup.

I like to either be entirely data driven, so that the input and output data are stored as data in files together so that it's obvious that this input leads to that output, which doesn't work here and would be overkill, or put the test setup data in the same file and ideally the same function as the tests so at least it's easy to see both at the same time.

As such, I'd move all of this except `set_uid`, `set_gid`, and maybe for creating `/home` into the test file rather than putting it in `conftest.py` and restructure this so that the setup is done inline immediately before the test that uses it. In other words, make the data for leni and then do the tests. Make the data for karl and then do those tests. All of that can be inline directly in the test, so then it becomes very obvious what the setup is and then what the test is checking. The test can just hard-code all of the paths and UIDs and GIDs that way rather than setting up a bunch of data structures that are only used in one place.

File: inithome/src/rubin/nublado/inithome/provisioner.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+
+def main() -> None:
+    """Entry point for provisioner.
+
+    Environment variables `NUBLADO_UID` and `NUBLADO_HOME` must be set.
+    if NUBLADO_GID is unset it will get set to NUBLADO_UID.
+    """
+    uid = int(environ["NUBLADO_UID"])  # Fail if unset
+    gid = int(getenv("NUBLADO_GID", uid))  # Default to UID
```
Comment (rra) [inline inithome/src/rubin/nublado/inithome/provisioner.py:96]: GID is now mandatory everywhere in the controller, so this fallback can be removed.

File: inithome/pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+[project.optional-dependencies]
+dev = [
+    # Testing
+    "pytest",
+    "pytest-cov",
+    "pytest-asyncio",
+    "pyfakefs",
+    # To work with higher-level nox/pyproject
+    "pydantic"
+]
```
Comment (rra) [inline inithome/pyproject.toml:32]: Since the output here is a Docker container rather than something that's pip-installable, I think this should use frozen dependencies like a service, and thus `requirements/dev.in` and so forth. It's total overkill, I know, but it maintains consistency between the controller and inithome so that we don't introduce yet a third pattern that's a hybrid between the pip-installable plugins and the things that generate a Docker file.

I assume "To work with higher-level nox/pyproject" is because the top-level pyproject.toml configures the Pydantic mypy plugin and therefore Pydantic has to be installed for mypy to run? If so, I'd say that explicitly; right now it's not clear what this is referring to.

**Review by rra — COMMENTED** [review]
The pull request title is now out of date, and the pull request could probably use a bit of a description.

File: inithome/src/rubin/nublado/inithome/provisioner.py (hunk 2 above)
Comment (athornton) [reply inithome/src/rubin/nublado/inithome/provisioner.py:29]: Oh, I can't either.  I just did that yesterday afternoon because I was peeved at Merlin.

File: inithome/src/rubin/nublado/inithome/provisioner.py (hunk 4 above)
Comment (athornton) [reply inithome/src/rubin/nublado/inithome/provisioner.py:69]: Per discussion let's just fail if the entire parent path is not present.

File: inithome/src/rubin/nublado/inithome/provisioner.py (hunk 5 above)
Comment (athornton) [reply inithome/src/rubin/nublado/inithome/provisioner.py:78]: Yeah, let's just insist that the parent already be there.

File: inithome/pyproject.toml (hunk 1 above)
Comment (athornton) [reply inithome/pyproject.toml:3]: Eh, I'd prefer to keep it inside an obviously-ours namespace.

File: inithome/pyproject.toml (hunk 9 above)
Comment (athornton) [reply inithome/pyproject.toml:32]: Easy stuff hammered out.

File: inithome/LICENSE
[the remark this answers is not in the document]
Comment (athornton) [reply inithome/LICENSE:1]: moved

File: inithome/pyproject.toml (hunk 9 above)
Comment (athornton) [reply inithome/pyproject.toml:32]: requirements updated.

File: inithome/src/rubin/nublado/inithome/provisioner.py (hunk 3 above)
Comment (athornton) [reply inithome/src/rubin/nublado/inithome/provisioner.py]: Done.

**Comment by athornton** [comment]
Belay.  Bad force-push.  Gotta go into the office and recover (well, probably don't *gotta* but that's going to be easiest)

**Comment by athornton** [comment]
I have de-hosed it.  Ready for review again.

File: inithome/requirements/dev.in [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+pytest-asyncio
+pytest-cov
+pytest-sugar
+respx
+types-pyyaml
+
+# Documentation
+autodoc_pydantic
+documenteer[guide]>=1.0.0a7
+sphinx-diagrams
```
Comment (rra) [inline inithome/requirements/dev.in:37]: Looks like some left-over cut and paste that can be removed.

File: inithome/src/rubin/nublado/inithome/provisioner.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+            if (stat_results.st_mode & 0o777) != 0o700:
+                logger = logging.getLogger(__name__)
+                logger.warning(
+                    f"{self.homedir} has strange permissions "
+                    f"{oct(stat_results.st_mode & 0o777)} rather than '0o700'"
+                )
+            return
+        # We need to create the directory
+        self.homedir.mkdir()
+        self.homedir.chmod(mode=0o700)
```
Comment (rra) [inline inithome/src/rubin/nublado/inithome/provisioner.py:78]: Minor security nit, but umask and then mkdir is better than mkdir and then chmod. If you somehow get run with a umask of 000, you create a world-writable home directory briefly and create a security race condition. Since you know what permissions you want, you can just set umask explicitly to avoid the race.

**Review by rra — COMMENTED** [review]
Just a couple of minor nits; otherwise, looks good. Thank you!

File: inithome/requirements/dev.in (hunk 10 above)
Comment (athornton) [reply inithome/requirements/dev.in]: Removed.
